### PR TITLE
snapstate: fix autorefresh from classic->strict

### PR DESCRIPTION
**Pull request by mvo5**
We got a bugreport that going from classic->strict in a refresh
does not work. This commit fixes this issue.

Back in c6117368b75 code was added in validateFlagsForInfo() to
check that it is not possible to do:
```
$ snap install --classic test-snapd-tools
```
i.e. this resulted in an error with this check.

However in 9e3b3a43f32 this was changed so that it is now possible
to do such an install. If this happens the --classic option is
simply ignored.

This commit now changes the code from 9e3b3a43f32 to also apply
to refreshes, not just installs.  This fixes refreshes that go
from classic->strict. This unblocks multipass (LP: #1883538).

I renamed the checkInstallPreconditions to ensureInstallPreconditions because
this code is now doing more than just checking.

### CODE REVIEW

**Comment by pedronis** [comment]
the changes look ok but we need to make sure we also have unit tests for this. Afaict there was already almost a fix in the codebase:

https://github.com/snapcore/snapd/blob/master/overlord/snapstate/snapstate.go#L947

spot the issue?

And there is code for the single snap case:

https://github.com/snapcore/snapd/blob/master/overlord/snapstate/snapstate.go#L1490

we need to drop that code if we have the new one, but make sure all of these have unit tests.

File: overlord/snapstate/snapstate.go [hunk 1]
```diff
@@ -597,8 +597,13 @@ func validateFeatureFlags(st *state.State, info *snap.Info) error {
 	return nil
 }
 
-func checkInstallPreconditions(st *state.State, info *snap.Info, flags Flags, snapst *SnapState, deviceCtx DeviceContext) error {
-	if err := validateInfoAndFlags(info, snapst, flags); err != nil {
+func ensureInstallPreconditions(st *state.State, info *snap.Info, flags *Flags, snapst *SnapState, deviceCtx DeviceContext) error {
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:600]: my slight preference would be for this to return the mutated flags instead of mutating in place

Comment (pedronis) [reply overlord/snapstate/snapstate.go:600]: actually not slight, if ones passes &snapst.Flags we could get behavior that is not quite right vs undos

Comment (pedronis) [reply overlord/snapstate/snapstate.go:600]: mmh, likely not undo, but we do some delicate things with snapst in snapstate.go that don't expect it to change under their back, so really prefer to be careful here

File: tests/main/refresh-classic/task.yaml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+prepare: |
+    #shellcheck source=tests/lib/dirs.sh
+    . "$TESTSLIB"/dirs.sh
+
+    case "$SPREAD_SYSTEM" in
+        fedora-*|arch-*|centos-*)
+            # although classic snaps do not work out of the box on fedora,
+            # we still want to verify if the basics do work if the user
+            # symlinks /snap to $SNAP_MOUNT_DIR themselves
+            ln -sf $SNAP_MOUNT_DIR /snap
```
Comment (zyga) [inline tests/main/refresh-classic/task.yaml:15]: I didn't check this deeply but IIRC we provide this symlink in global prepare. This code doesn't check for absence of the symlink. The restore section should be conditional on the initial absence of the symlink

Comment (mvo5) [reply tests/main/refresh-classic/task.yaml:15]: I just double checked in spread:
```
google:fedora-31-64 .../tests/main/refresh-classic# ls -l /snap
ls: cannot access '/snap': No such file or directory
```

Comment (zyga) [reply tests/main/refresh-classic/task.yaml:15]: Thanks for checking. I think I would still prefer to be more careful here. If you don't mind I can either change the code in this PR or we can merge it and I'll send a follow-up.

Comment (mvo5) [reply tests/main/refresh-classic/task.yaml:15]: Thanks, this pattern is used in a bunch of other tests too, I cargo culted it :( Let's do a followup that move this into some sort of helper where we can then DTRT in a single place.

**Pull request merged**